armory/scenarios/librispeech_dev_clean_spectrogram.py

```diff
@@ -0,0 +1,171 @@
+
+        task_metric = metrics.categorical_accuracy
+
+        # Train ART classifier
+        if not model_config["weights_file"]:
+            classifier.set_learning_phase(True)
+            logger.info(
+                f"Fitting model {model_config['module']}.{model_config['name']}..."
+            )
+            fit_kwargs = model_config["fit_kwargs"]
+            train_data_generator = load_dataset(
+                config["dataset"],
+                epochs=fit_kwargs["nb_epochs"],
+                split_type="train",
+                preprocessing_fn=preprocessing_fn,
+            )
+
+            for cnt, (x, y) in tqdm(enumerate(train_data_generator)):
+                # x is of shape (N,241,T), representing N spectrograms,
+                # each with 241 frequency bins and T time bins that's variable,
+                # depending on the duration of the corresponding raw audio.
+                # The model accepts a fixed size spectrogram, so x_trains need to
+                # be sampled.
+
+                x_seg = []
+                for xt in x:
+                    rand_t = np.random.randint(xt.shape[1] - n_tbins)
+                    x_seg.append(xt[:, rand_t : rand_t + n_tbins])
+                x_seg = np.array(x_seg)
+                x_seg = np.expand_dims(x_seg, -1)
+
+                classifier.fit(
+                    x_seg,
+                    y,
+                    batch_size=config["dataset"]["batch_size"],
+                    nb_epochs=1,
+                    verbose=True,
+                )
+
+                if cnt % train_data_generator.batches_per_epoch == 0:
+                    # evaluate on validation examples
+                    val_data_generator = load_dataset(
+                        config["dataset"],
+                        epochs=1,
+                        split_type="validation",
+                        preprocessing_fn=preprocessing_fn,
+                    )
+
+                    cnt = 0
+                    validation_accuracies = []
+                    for x_val, y_val in tqdm(val_data_generator):
+                        x_val_seg, y_val_seg = [], []
+                        for xt, yt in zip(x_val, y_val):
+                            n_seg = int(xt.shape[1] / n_tbins)
+                            xt = xt[:, : n_seg * n_tbins]
+                            for ii in range(n_seg):
+                                x_val_seg.append(
+                                    xt[:, ii * n_tbins : (ii + 1) * n_tbins]
+                                )
+                                y_val_seg.append(yt)
+                        x_val_seg = np.array(x_val_seg)
+                        x_val_seg = np.expand_dims(x_val_seg, -1)
+                        y_val_seg = np.array(y_val_seg)
+
+                        y_pred = classifier.predict(x_val_seg)
+                        validation_accuracies.extend(task_metric(y_val_seg, y_pred))
+                        cnt += len(y_val_seg)
```

Comment:
I'm confused. Why are the validation and test labels segmented, but the training labels are not?

Reply:
Is this because inference can take arbitrary length inputs but training cannot?

Reply:
No I didn't find a good reason this this. Instead of sampling all of the inputs it only sampled a single one. Updated.